File: 0045_Jump_Game_II.java
class Solution {
    public int jump(int[] nums) {
        int current = 0;
        int jumps = 0;
        
        while(current < nums.length - 1) {
            if(nums[current] + current >= nums.length - 1) return ++jumps;    
            
            // find the reachable cell with the farthest reach
            int farthestReacher = 0;
            int maxReach = 0;
            for(int i = current + 1; i <= current + nums[current]; i++) {
                int currentReach = i + nums[i];
                if(currentReach >= maxReach) {
                    maxReach = currentReach;
                    farthestReacher = i;
                }
            }
            
            current = farthestReacher;
            jumps++;
        }
        
        return jumps;
    }
}

File: 0045_Jump_Game_II.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        constexpr int MAX_LEN = 10'000;
        int dp[MAX_LEN]; // dp[i] holds the minimum number of steps to reach the last element
        
        dp[nums.size() - 1] = 0;
        
        for(int i = nums.size() - 2; i >= 0; i--) {
            // calculate dp[i]
            int best_jump = MAX_LEN + 1; // minimum number of moves to get to last elem after one jump
            for(int j = i + 1; j < nums.size() && j <= i + nums[i]; j++) 
                best_jump = min(best_jump, dp[j]);
            dp[i] = best_jump + 1;
        }
        
        return dp[0];
    }
};
